tcms/xmlrpc/api/testcase.py

```diff
@@ -879,20 +877,25 @@ def link_plan(request, case_ids, plan_ids):
             err_msg = 'TestPlan %s does not exist.' % ids_str
         raise ObjectDoesNotExist(err_msg)
 
+    # (plan_id, case_id) pair might probably exist in test_case_plans table, so
+    # skip the ones that do exist and create the rest.
+    # note: this query returns a list of tuples!
+    existing = TestCasePlan.objects.filter(
+        plan__in=plan_ids,
+        case__in=case_ids
+    ).values_list('plan', 'case')
+
     # Link the plans to cases
     def _generate_link_plan_value():
         for plan_id in plan_ids:
             for case_id in case_ids:
-                yield plan_id, case_id
+                if (plan_id, case_id) not in existing:
```

Comment:
There is a unique constraint in table to avoid inserting duplicate pair of plan and case. This check could be a good way, but it changes original behavior of this API. Currently, I would not prefer to change original behavior of any API even if it's really good to do that. All of them the good things should go to new version of API whatever XMLRPC API or new RESTful API.

BTW, are you using XMLRPC API now?